### ISSUE TITLE
sending special keys is broken on Samsung Galaxy Tab S6 with matching hardware keyboard

### ISSUE THREAD

**Issue by bk138** [bk138/multivnc#30]

**Comment by TT44bb** [comment]
Would really appreciate a fix. Often not even enter works because it focuses on some multivnc ui element.

**Comment by bk138** [comment]
@TT44bb can you provide some more details or a way to reproduce? That'd be very helpful! :-)

**Comment by TT44bb** [comment]
I can't reproduce everything, but one thing I can reproduce is key combinations involving Alt L. When watching the input in xev, for the Alt-2 combo for example, there is a key down event for Alt, then one for 2, then there is a release event for 2, and then nothing - the key release event for alt is missing. This matches up with alt-key combos producing strange results. The Alt key on its own produces correct events, one down and one up for Alt L.

The other easily reproducible thing is everything involving Alt Gr. Pressing Alt Gr on its own causes key down events for both Alt L and Alt R, and corresponding up events.

Pressing the button that outputs < in Termux causes a down event for shift l, a down event for +, and up event for shift l and an up event for =.

Pressing the button that outputs - in Termux (or this text field here) causes a / down and up event.

Also there is a special button in the keyboard that doesn't produce any events, labelled "Lang (fn alternative: show/hide touch keyboard)". This probably works in a non-standard way, but it would be great if it was possible to map it to the super key (because that in turn is missing from the keyboard).

All of this with multivnc from f-droid, connecting to a Debian chroot that was created by AnLinux which is also in f-droid and really straightforward to use. Hardware is a Samsung Galaxy Tab S6.

**Comment by bk138** [comment]
@TT44bb what kind of keyboard (hard/soft) are you using?

**Comment by TT44bb** [comment]
I just got an update for the keyboard (who makes keyboards with software updates?) and now the key events are broken in a different way (maybe slightly less).

I feel this bug should be put on hold until I test with a different keyboard. Fucking Samsung selling $800 tablets with broken keyboards.

**Comment by sudomain** [comment]
I'm experiencing a similar issue when using MultiVNC to connect to an Arch Linux proot in Termux. I've tested the following with both a Nulaxy KM12 bluetooth keyboard as well as the Hacker's Keyboard app from F-Droid:
* When typing commands into XTerm, the Alt key is mostly ignored for shortcuts that should normally work (Alt + d should delete sequences of non-whitespace characters aka "words", Alt + t should swap words). Other alt-key combinations insert strange characters such as � (produced with alt +S)

* Ctrl key shortcuts work, but not Ctrl + Shift shortcuts. MultiVNC will correctly send a shortcut such as Ctrl + A, but if a shortcut such as Ctrl + Shift + B is typed, then Ctrl + B will be sent instead

**Comment by lukas-buergi** [comment]
> I'm experiencing a similar issue when using MultiVNC to connect to an Arch Linux proot in Termux. I've tested the following with both a Nulaxy KM12 bluetooth keyboard as well as the Hacker's Keyboard app from F-Droid

This sounds like you are using a different device, which doesn't come with its own hardware keyboard. Are you? Or are you using a Samsung Tab S6 and didn't get the matching hardware keyboard?

Because the way it's broken on my Samsung device makes me believe that Samsung fucked it up and multivnc is working. Especially because none of the other VNC apps that I tried worked either. I looked a little into Android input event handling, but haven't yet gotten around to testing it more thoroughly. My tablet became an overly expensive multi media device instead of the workhorse I expected.

**Comment by sudomain** [comment]
> This sounds like you are using a different device, which doesn't come with its own hardware keyboard. Are you? Or are you using a Samsung Tab S6 and didn't get the matching hardware keyboard?

I'm not using a S6 nor a hardware keyboard that came with my device, but a generic bluetooth keyboard. I (maybe wrongly) assumed that since this issue is still open, that it is about how MultiVNC sends key combinations/shortcuts on different keyboards rather than a specific hardware issue. I thought I'd share my similar problems with the `Alt ` key and also the problem of `Ctrl + Shift ` shortcuts. If a separate issue should be made for my specific device, then I'll do that.

**Comment by lukas-buergi** [comment]
> > This sounds like you are using a different device, which doesn't come with its own hardware keyboard. Are you? Or are you using a Samsung Tab S6 and didn't get the matching hardware keyboard?
> 
> I'm not using a S6 nor a hardware keyboard that came with my device, but a generic bluetooth keyboard. I (maybe wrongly) assumed that since this issue is still open, that it is about how MultiVNC sends key combinations/shortcuts on different keyboards rather than a specific hardware issue. I thought I'd share my similar problems with the `Alt ` key and also the problem of `Ctrl + Shift ` shortcuts. If a separate issue should be made for my specific device, then I'll do that.

I think collecting the problems here until someone verifies whether it's a multivnc bug or a device software bug is fine.

**Comment by sudomain** [comment]
Apologies/Update: the `Alt + (letter)` issue that I described is likely an issue with the XTerm package I was using. After testing a terminal emulator with more features (Konsole) it appears that `Alt + (letter)` shortcuts do in fact get passed through properly, though this test revealed another slight issue: My keyboard has two Alt keys, but the `Alt + (letter)` shortcuts only work when using the Alt key on the right side of the keyboard. The left Alt key appears to be ignored by MultiVNC (`Alt + T` just passes the letter `T` through. On VNC Viewer however, both Alt keys work.

Another thing I tested: the `Ctrl + Shift + (letter)` shortcuts work in the VNC Viewer app, though not on MultiVNC.
